### PR TITLE
Require secure renegotiation support by default

### PR DESCRIPTION
**Pull request by kaduk**
Per #14848 , stop setting `SSL_OP_LEGACY_SERVER_CONNECT` by default, and don't include it in `SSL_OP_ALL`.
Deal with fallout in docs and tests (including fixing a bug in `SSL_CONF` that only lets you set the flag for servers, when the flag is only useful on clients!).

@davidben please remind me if there were any issues with other implementations thinking that advertising renegotiation-info support implies a willingness to support renegotiation at all, that we might need to consider.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by richsalz** [comment]
hooray for this, further 'secure by default'

File: test/bad_dtls_test.c [hunk 1]
```diff
@@ -494,6 +494,8 @@ static int test_bad_dtls(void)
     if (!TEST_ptr(ctx)
             || !TEST_true(SSL_CTX_set_min_proto_version(ctx, DTLS1_BAD_VER))
             || !TEST_true(SSL_CTX_set_max_proto_version(ctx, DTLS1_BAD_VER))
+            || !TEST_true(SSL_CTX_set_options(ctx,
+                                              SSL_OP_LEGACY_SERVER_CONNECT))
```
Comment (mattcaswell) [inline test/bad_dtls_test.c:498]: Why is this necessary?

Comment (kaduk) [reply test/bad_dtls_test.c:498]: I only got as far as "the test failed if it wasn't there".
I guess I can try to dig into it some more...

Comment (kaduk) [reply test/bad_dtls_test.c:498]: Ah, it seems that bad_dtls_test eschews our DTLS server stack and just manually does the minimal message processing possible, which includes `static unsigned char server_hello[] = {...` that is obviously going to make a client that expects `renegotiation_info` sad.
Perhaps @dwmw2 has some insight into whether actual servers support renegotiation info or not, which might influence whether we leave this flag set or adapt the server portion of the test.

Comment (dwmw2) [reply test/bad_dtls_test.c:498]: > Ah, it seems that bad_dtls_test eschews our DTLS server stack and just manually does the minimal message processing possible

Indeed so. I don't believe we support Cisco's bizarre pre-DTLS-1.0 variant of the protocol in server mode at all, and even if we did I wouldn't necessarily want bad_dtls_test to test against it. Testing our own code against *itself* isn't quite the interop test I had in mind :)

> Perhaps @dwmw2 has some insight into whether actual servers support renegotiation info or not, which might influence whether we leave this flag set or adapt the server portion of the test.

No, I'm fairly convinced that the Cisco ASA is never going to grok renegotiation_info for DTLS_BAD_VER; turning it and all other options off is the right thing to do there.

Comment (kaduk) [reply test/bad_dtls_test.c:498]: > No, I'm fairly convinced that the Cisco ASA is never going to grok renegotiation_info for DTLS_BAD_VER; turning it and all other options off is the right thing to do there.

Thanks; that's exactly the type of data I was looking for!

Comment (mattcaswell) [reply test/bad_dtls_test.c:498]: In that case this PR is likely to be a breaking change for users of DTLS_BAD_VER. It's still the right thing to do though.

Comment (dwmw2) [reply test/bad_dtls_test.c:498]: I guess I should use this in OpenConnect too, something like this:
https://gitlab.com/openconnect/openconnect/-/commit/f3f024f587b564c7a80818426793a1f3b6119833

What version of OpenSSL is this expected to land in, and will there be an API soname bump that correlates to it?
In particular, might an application be built against an older version of OpenSSL, but *run* against a version which requires this flag? That would give a *false* binary compatibility... and would mean that I should define `SSL_OP_LEGACY_SERVER_CONNECT` for myself in my own code even if it isn't present in OpenSSL headers at build time, and always *try* it at run time just in case?

**Review by mattcaswell — APPROVED** [review]

File: test/bad_dtls_test.c (hunk 1 above)
Comment (mattcaswell) [reply test/bad_dtls_test.c:498]: This is 3.0 only thing - so its a major version change.

Comment (kaduk) [reply test/bad_dtls_test.c:498]: I expect this to be in 3.0.0 which is a major release and has a soname bump.
It is not supposed to be possible to have a build/runtime mismatch in this case.

Note that `SSL_OP_LEGACY_SERVER_CONNECT` has been around in the code for a decade, and this change just affects the default setting for it.  It is always safe for your code to always set or always clear it.

Comment (dwmw2) [reply test/bad_dtls_test.c:498]: > In that case this PR is likely to be a breaking change for users of DTLS_BAD_VER. It's still the right thing to do though.

Agreed. OpenConnect and potentially Cisco's own client are the only users of DTLS_BAD_VER, and my bad_dtls_test did its job perfectly here by catching the breakage in good time for me to cope.

Comment (dwmw2) [reply test/bad_dtls_test.c:498]: Confirming that my patch above works for DTLS connections, but I also can't even connect to the **HTTPS** service of the Cisco ASA without setting `SSL_OP_LEGACY_SERVER_CONNECT`. Yay Cisco!

```
SSL negotiation with vpn.intel.com
SSL connection failure
C007DC46F77F0000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:ssl/statem/extensions.c:880:
Failed to open HTTPS connection to vpn.intel.com
```

**Comment by kaduk** [comment]
Merged to master; closing.
Thanks for the review!
